solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java

```diff
@@ -372,7 +372,7 @@ public void doc(int docId, StoredFieldVisitor visitor) throws IOException {
       Document cached = doc(docId);
       visitFromCached(cached, visitor);
     } else {
-      searcher.getIndexReader().document(docId, visitor);
+      searcher.getIndexReader().storedFields().document(docId, visitor);
```

Comment:
I wonder if `searcher.getIndexReader().storedFields()` can be constructed in the constructor? the searcher is stored there. This would avoid recreating in a few places in this file.

Reply:
I actually don't think this will work. SolrDocumentFetcher is a singleton instance variable in SolrIndexSearcher, and the analogous previous calls were thread-safe by virtue of ThreadLocal. StoredField instances are not thread-safe.

Reply:
I think this pattern -- calling storedFields() only to use it for exactly one document should be banned from any hot code path.  And SolrDocumentFetcher is such.  I sympathize that a more elegant solution isn't obvious but we should do something.  

Perhaps we change the contract of SolrIndexSearcher.getDocFetcher (which I added years ago BTW) to return a new instance (with a storedFields of course) that is not thread-safe.  But nonetheless has all the existing caches of any other instance pulled from the same searcher.  Hmmmm.

Or use an existing well-managed ThreadLocal -- SolrRequestInfo, to hold the StoredField instance.  If it's null (unexpected / non-standard use cases) then the fallback is grab a storedFields() one-off use.  I think this would work well.

Reply:
If a performance benchmark returning, say, 1000 docs (not big docs) is not impacted, I suppose the change as-is is fine albeit not ideal.

Reply:
> If a performance benchmark returning, say, 1000 docs (not big docs) is not impacted, I suppose the change as-is is fine albeit not ideal.

Just noting that I myself have no bandwidth/plans to work on that i.e. it's up-for-grabs opportunity for someone else. Thanks!

Reply:
> Perhaps we change the contract of SolrIndexSearcher.getDocFetcher (which I added years ago BTW) to return a new instance (with a storedFields of course) that is not thread-safe.

75e095552e51302096883a96a93cf9197140f3a3 implements this approach. Feedback welcome; feel free to revert or rework as you see fit.